src/network/network_gui.cpp

```diff
@@ -2514,3 +2517,131 @@ void ShowNetworkAskRelay(const std::string &server_connection_string, const std:
+		}
+	}
+
+	void DrawWidget(const Rect &r, int widget) const override
+	{
+		if (widget == WID_NAS_TEXT) {
+			DrawStringMultiLine(r.Shrink(WidgetDimensions::scaled.frametext), STR_NETWORK_ASK_SURVEY_TEXT, TC_FROMSTRING, SA_CENTER);
+		}
+	}
+
+	void FindWindowPlacementAndResize(int def_width, int def_height) override
+	{
+		/* Position query window over the calling window, ensuring it's within screen bounds. */
+		this->left = Clamp(parent->left + (parent->width / 2) - (this->width / 2), 0, _screen.width - this->width);
+		this->top = Clamp(parent->top + (parent->height / 2) - (this->height / 2), 0, _screen.height - this->height);
+		this->SetDirty();
+	}
+
+	void OnClick(Point pt, int widget, int click_count) override
+	{
+		switch (widget) {
+			case WID_NAS_NO:
+				_settings_client.network.participate_survey = PS_NO;
+				this->Close();
+				break;
+
+			case WID_NAS_YES:
+				_settings_client.network.participate_survey = PS_YES;
+				this->Close();
+				break;
+		}
+	}
+};
+
+static const NWidgetPart _nested_network_ask_survey_widgets[] = {
+	NWidget(NWID_HORIZONTAL),
+		NWidget(WWT_CLOSEBOX, COLOUR_GREY),
+		NWidget(WWT_CAPTION, COLOUR_GREY, WID_NAS_CAPTION), SetDataTip(STR_NETWORK_ASK_SURVEY_CAPTION, STR_NULL),
+	EndContainer(),
+	NWidget(WWT_PANEL, COLOUR_GREY), SetPIP(0, 0, 8),
+		NWidget(WWT_TEXT, COLOUR_GREY, WID_NAS_TEXT), SetAlignment(SA_HOR_CENTER), SetFill(1, 1),
+		NWidget(NWID_HORIZONTAL, NC_EQUALSIZE), SetPIP(10, 15, 10),
+			NWidget(WWT_PUSHTXTBTN, COLOUR_YELLOW, WID_NAS_NO), SetMinimalSize(71, 12), SetFill(1, 1), SetDataTip(STR_NETWORK_ASK_SURVEY_NO, STR_NULL),
+			NWidget(WWT_PUSHTXTBTN, COLOUR_YELLOW, WID_NAS_YES), SetMinimalSize(71, 12), SetFill(1, 1), SetDataTip(STR_NETWORK_ASK_SURVEY_YES, STR_NULL),
+		EndContainer(),
+	EndContainer(),
+};
+
+static WindowDesc _network_ask_survey_desc(
+	WDP_CENTER, nullptr, 0, 0,
+	WC_NETWORK_ASK_SURVEY, WC_NONE,
+	WDF_MODAL,
+	_nested_network_ask_survey_widgets, lengthof(_nested_network_ask_survey_widgets)
+);
+
+/**
+ * Show a modal confirmation window with "no" / "yes" buttons.
+ */
+void ShowNetworkAskSurvey()
+{
+	/* If we can't send a survey, don't ask the question. */
+	if constexpr (!NetworkSurveyHandler::IsSurveyPossible()) return;
+
+	CloseWindowByClass(WC_NETWORK_ASK_SURVEY);
+
+	Window *parent = GetMainWindow();
+	new NetworkAskSurveyWindow(&_network_ask_survey_desc, parent);
+}
+
+/** Window for displaying the textfile of a survey result. */
+struct SurveyResultTextfileWindow : public TextfileWindow {
+	const GRFConfig *grf_config; ///< View the textfile of this GRFConfig.
+
+	SurveyResultTextfileWindow(TextfileType file_type) : TextfileWindow(file_type)
+	{
+		this->lines.clear();
+		auto result = _survey.CreatePayload(NetworkSurveyHandler::Reason::PREVIEW, true);
+
+		/* Put string in C-style string. */
+		this->text = ReallocT(this->text, result.size() + 1);
+		memcpy(this->text, result.c_str(), result.size());
+		this->text[result.size()] = '\0';
```

Comment:
Not that I'm fond of it... but... why not free and stredup? Now you're essentially hiding the C-style string allocation.
But alas... #10772 would make this whole problem go away I reckon.

Reply:
There is a reason I started #10772, yes :) But it depends on #10776 .. and down the rabbithole we go :) Will be fine before this gets undrafted :)